internal/state/match.go

```diff
@@ -21,52 +21,78 @@ type Match struct {
 func getMatches(
 	ctx context.Context,
 	client *ent.Client,
-	mailboxes []*ent.Mailbox,
+	allMailboxes []*ent.Mailbox,
 	ref, pattern, delimiter string,
 	subscribed bool,
 ) (map[string]Match, error) {
 	matches := make(map[string]Match)
 
-	for _, mailbox := range mailboxes {
-		if subscribed && !mailbox.Subscribed {
-			continue
-		}
+	mailboxes := make(map[string]*ent.Mailbox)
+	for _, mbox := range allMailboxes {
+		mailboxes[mbox.Name] = mbox
+	}
 
+	for mboxName := range mailboxes {
 		select {
 		case <-ctx.Done():
 			return nil, ctx.Err()
 
 		default: // fallthrough
 		}
 
-		if name, ok := match(ref, pattern, delimiter, mailbox.Name); ok {
-			if mailbox.Name == name {
-				atts := imap.NewFlagSetFromSlice(xslices.Map(mailbox.Edges.Attributes, func(flag *ent.MailboxAttr) string {
-					return flag.Value
-				}))
+		for _, superior := range append(listSuperiors(mboxName, delimiter), mboxName) {
+			matchedName, ok := match(ref, pattern, delimiter, superior)
+			if !ok {
+				continue
+			}
 
-				recent, err := db.GetMailboxRecentCount(ctx, client, mailbox)
-				if err != nil {
-					return nil, err
-				}
+			if _, ok := matches[matchedName]; ok {
+				continue
+			}
 
-				if recent > 0 {
-					atts = atts.Add(imap.AttrMarked)
-				} else {
-					atts = atts.Add(imap.AttrUnmarked)
-				}
+			mbox, ok := mailboxes[matchedName]
 
-				matches[mailbox.Name] = Match{
-					Name:      mailbox.Name,
-					Delimiter: delimiter,
-					Atts:      atts,
-				}
-			} else {
-				matches[name] = Match{
-					Name:      name,
+			// skip match when:
+			if subscribed && !mbox.Subscribed && // should be subscribed and it's not
+				(matchedName == mboxName || !strings.HasSuffix(pattern, "%")) { // is not superior or percent wildcard not used
+				continue
+			}
+
+			// add match as NoSelect when:
+			if !ok || // is deleted superior
+				matchedName == "" || // is empty request for delimiter response
+				subscribed && !mbox.Subscribed { // is unsubscribed superior
+				matches[matchedName] = Match{
+					Name:      matchedName,
 					Delimiter: delimiter,
 					Atts:      imap.NewFlagSet(imap.AttrNoSelect),
 				}
+
+				continue
```

Comment:
could you extract this stuff into some helper functions `skipMatch` and `addMatch`?